tests/kernel/smp/src/main.c

```diff
@@ -37,45 +52,354 @@ void t2_fn(void *a, void *b, void *c)
+ * lower priority threads. So higher priority thread should occupy
+ * one core and lower priority threads should run in round
+ * robin.
+ */
+static void test_smp_single_core_preemp_rr_threads(void)
+{
+	int i;
+
+	k_thread_priority_set(k_current_get(),
+			      K_PRIO_PREEMPT(K_HIGHEST_THREAD_PRIO));
+
+	k_sched_time_slice_set(100, 0);
+
+	for (i = 0; i < THREADS_NUM; i++) {
+		tdata[i].executed = 0;
+		tdata[i].tid = k_thread_create(&tthread[i], tstack[i],
+					       STACK_SIZE, smp_thread_entry,
+					       (void *)i, NULL, NULL,
+					       K_PRIO_PREEMPT(
+						       K_LOWEST_THREAD_PRIO - 1),
+					       0, 0);
+	}
+
+	k_busy_wait(THREAD_DELAY * THREADS_NUM + 10);
+
+	for (i = 0; i < THREADS_NUM; i++) {
+		k_thread_abort(tdata[i].tid);
+	}
+
+	/* Test point: All THREADS_NUM spawned threads must have
+	 * executed in another core
+	 */
+	for (i = 0; i < THREADS_NUM; i++) {
+		zassert_true(tdata[i].executed == 1, "Threads didn't get"
+			     " scheduled on another core");
+	}
+}
+
+/**
+ * @brief Verify round robin on multi core with coop thread
+ *
+ * @details A cooperative thread creates THREAD_NUM preemptive
+ * threads of equal priority. After creation, cooperative thread
+ * goes for sleep till all preemptive threads finish execution.
+ * It checks if all threads are executed and returns test execution
+ * results
+ */
+static void test_smp_multi_core_coop_rr_threads(void)
+{
+	int i;
+
+	k_sched_time_slice_set(100, 0);
+
+	for (int i = 0; i < THREADS_NUM; i++) {
+		tdata[i].executed = 0;
+		tdata[i].tid  = k_thread_create(&tthread[i], tstack[i],
+						STACK_SIZE, smp_thread_entry,
+						(void *)i, NULL, NULL,
+						K_PRIO_PREEMPT(
+							K_LOWEST_THREAD_PRIO - 1), 0, 0);
+	}
+	/* THREAD_DELAY is the parameter to k_busy_wait(). So THREAD_DELAY
+	 * is in useconds. And k_sleep()'s parameter is in milliseconds.
+	 */
+	k_sleep(((THREAD_DELAY / 1000) * THREADS_NUM)
+		/ CONFIG_MP_NUM_CPUS + 100);
+
+	for (i = 0; i < THREADS_NUM; i++) {
+		k_thread_abort(tdata[i].tid);
+	}
+	/* Test point: All THREADS_NUM spawned threads must have
+	 * executed.
+	 */
+	for (i = 0; i < THREADS_NUM; i++) {
+		zassert_true(tdata[i].executed == 1, "Threads didn't get"
+			     " scheduled on all the cores");
+	}
+}
+
+/**
+ * @brief Verify round robin on multiple cores with preempt threads
+ *
+ * @details The test basically validates for the scenario
+ * where parent has to execute after all child processes
+ * execution is complete
+ */
+static void test_smp_multi_core_preemp_rr_threads(void)
+{
+	int i;
+
+	k_thread_priority_set(k_current_get(),
+			      K_PRIO_PREEMPT(K_LOWEST_THREAD_PRIO));
+
+	k_sched_time_slice_set(100, 0);
+
+	for (int i = 0; i < THREADS_NUM; i++) {
+		tdata[i].executed = 0;
+		tdata[i].tid = k_thread_create(&tthread[i], tstack[i],
+					       STACK_SIZE, smp_thread_entry,
+					       (void *)i, NULL, NULL,
+					       K_PRIO_PREEMPT(
+						       K_LOWEST_THREAD_PRIO - 1),
+					       0, 0);
+	}
```

Comment:
Yet more "spawn all the threads" boilerplate that could be unified.  Also K_LOWEST_APPLICATION_THREAD_PRIO.